config/registry.yaml

```diff
@@ -12,3 +12,13 @@ registry:
     collections: lsst.daf.butler.registry.collections.synthIntKey.SynthIntKeyCollectionManager
     datasets: lsst.daf.butler.registry.datasets.byDimensions.ByDimensionsDatasetRecordStorageManager
     datastores: lsst.daf.butler.registry.bridge.monolithic.MonolithicDatastoreRegistryBridgeManager
+  schema_versions:
+    core:
+      version: 0.0.1
+      digest: ""
+    dimensions:
+      version: 0.0.1
+      digest: ""
+    datastores:
```

Comment:
Another quick point, but in theory the schema for datastores can depend on the datastore. s3datastore and posixDatastore do use the same schema for their internal table but that's not required (inMemoryDatastore doesn't use a table at all).

Reply:
If the way this would work in practice is that the single datastore schema version would need to be updated whenever any datastore implementation changes its schema, that might still be okay (at least if all datastores continue to be implemented in daf_butler, not as extensions in other packages - that worries me a bit, but maybe that's just a hypothetical concern). It might cause us to declare some repos as incompatible when they actually aren't, which is better than not recognizing an incompatibility, but it could still be a problem.

Of course, we have the exact same problem (I think) with the core schema version and all of the registry managers.  If we change one of those, I assume that means we change the core schema version - but that implies that repos that didn't use the changed manager now report as incompatible with the codebase, but really aren't.

Reply:
Hmm, it gets complicated. I think that schema is not the whole story, there is also data compatibility. Could s3datastore and posixDatastore actually share the same database? If they cannot then we should not use the same version number for their schema.

That question also applies to other potential cases, I can imagine that sometimes we may need to do a data migration without changing schema (e.g. changing the format/meaning of string/BLOB data). That may also be accompanied by the change of the manager class but not necessarily. That sort of change also needs to be reflected in the schema version. And I think that configuration that we are going to store in the new attributes table is another example where the data in a table has to be consistent with the schema version number because that configuration actually determines the schema.

Reply:
S3 and Posix datastores do currently share the same database schema but nothing requires that they continue to do so.  It is defined in fileLikeDatastore at the moment.

Reply:
My question was more about data, suppose that Posix datastore had written something to that table, could S3 datastore read and interpret those records, or at least understand that they are from different datastore and have to be ignored?

Reply:
each datastore has its own table. They don't share content. You could point an S3 datastore at the posix table and it would work so long as somebody behind the scenes had uploaded the data to S3 without telling butler but that wasn't something we had considered being a standard scenario.

Reply:
And each butler can have multiple datastores in play each with their own tables and you can only work out how many you have by asking the datastore itself.

Reply:
If there can be multiple datastores and/or datastores can be added/removed dynamically then I think we cannot manage that kind of setup via a single version number. Each datastore probably needs its own separate schema version number and each one needs to me migrated separately. I think in datastores case we likely don't need to migrate database schema between different datastores, so they can be managed completely independently, each one of them would have a linear history. It would still be beneficial to have the same tool (alembic) to manage datastores versions, so those versions need to be identifiable like for other managers (via configuration or in the source code).

Reply:
I think each datastore class (s3, posix etc) needs its own version number.